pkg/webhook/resources/storageclass/validator.go

```diff
@@ -229,3 +233,15 @@ func (v *storageClassValidator) validateVMImageUsage(sc *storagev1.StorageClass)
 
 	return nil
 }
+
+// The `harvester-longhorn` SC is created from helm chart and monitored by the managedchart, also used by rancher-monitoring.
+// It should not be deleted accidentally.
+func (v *storageClassValidator) validateHarvesterLonghornSC(sc *storagev1.StorageClass) error {
+	if sc.Name == util.StorageClassHarvesterLonghorn && sc.Annotations != nil {
+		if sc.Annotations[util.HelmReleaseNamespaceAnnotation] == util.HarvesterSystemNamespaceName && sc.Annotations[util.HelmReleaseNameAnnotation] == util.HarvesterChartReleaseName {
+			return werror.NewInvalidError(fmt.Sprintf("storage class %s is reserved by Harvester and can't be deleted", sc.Name), "")
+		}
+	}
+
+	return nil
+}
```

Comment:
The whole annotations and labels are:

```yaml
metadata:
  annotations:
    meta.helm.sh/release-name: harvester
    meta.helm.sh/release-namespace: harvester-system
    objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
    storageclass.kubernetes.io/is-default-class: "true"
  labels:
    app.kubernetes.io/managed-by: Helm
    objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
```

Currently, users can't upgrade the Harvester due to losing `objectset.rio.cattle.io/id` and `objectset.rio.cattle.io/hash`. Besides checking helm installation, I suggest that we also need to check these two fields.

Reply:
The current code is assuming that certain labels/annotations are set by the helm & managedchart; it is not the best idea to list all the known labels/annotations here from following considerations:

(0) The harvester-longhorn is defined in Harvester chart https://github.com/harvester/harvester/blob/606f485579d087def560e2a8a255a2b0f4bad019/deploy/charts/harvester/templates/harvester-storageclass.yaml#L5

(1) Harvester supports developer mode, install from helm directly, https://docs.harvesterhci.io/v1.3/developer/developer-mode-installation

(2) The managedchart is on-top of helm.

The helm related annotations will be there; managedchart related may be there. For webhook, check the helm related annotations works for most scenarios.

An drawback is that, no one is suring those annotations won't change.

(3) A better solution is to add a harvester releated label/annotation in the chart definition, thus the webhook can drop the dependency on helm/managedchart. I will add this.

It may not work for legacy deployments.